### ISSUE TITLE
citation within caption crashes pandoc if csl specified

### ISSUE THREAD

**Issue by colcord** [jgm/pandoc-citeproc#188]
pandoc without csl expands citation within caption.
pandoc with csl crashes in various ways depending on csl

If I should report this as a bug against the csl, please let me know how. I've included in the text files the places I have found where I might file additional bug reports.

I attach two markdown documents.
One (CaptionTest5xkcd4.txt) shows the different error messages that pandoc generates with each csl file.
(Let me know if I should give URLs to the csl files I have used).

The other document (CaptionTest5xkcd6.txt) runs through pandoc, and shows the different things I have tried. This document is the same as the first with pandoc error messages taken out.

Thanks for all your hard work on pandoc. It is wonderful to see an open source project thriving. 
I hope I can do my bit to help make it better.

I can't attach text files, so I will copy the contents of CaptionTest5xkcd4.txt below.

I can't attach bib file, so this is the contents:

@online{_xkcd,
  title = {xkcd: {Wikipedian} {Protester}},
  url = {https://xkcd.com/285/},
  timestamp = {2015-10-30T07:53:50Z},
  urldate = {2015-10-30},
  note = {00000},
  file = {xkcd\: Wikipedian Protester:/home/frank/Zotero/storage/SM3Q5I5G/285.html:}
}

CaptionTest5xkcd4.txt

% Caption Citation Issues
% Frank
% November 2015

I have been helping a friend write a dissertation. 
Pandoc has been useful, producing good looking pdfs using a csl and template.

## Citation format
Her citation format needs Author Title Year in the footnotes. We are currently using modern-humanities-research-association.csl.

## Images with Citations

She needs many images in her dissertation which come from different sources.

There have been some issues with citations in captions.

This document is meant to be a way to test what causes the issues.
I have tried to make it self contained by using a web image.

I checked my version of pandoc, and it was out of date. Have updated to 1.15.1.
Cite-proc apparently updated at the same time.

![caption with only text, web image](http://imgs.xkcd.com/comics/wikipedian_protester.png)

The above image and caption should come out fine. It's a bit small with the xkcd web image.

This citation  @_xkcd seems fine in the bib file.
Bib file generated from Zotero and export items as  Betterbib latex format. The citation comes out fine in the pdf. 

The image above is a bit small. I thought Pandoc would automatically size it. I have a list of commands below which are trying to make it bigger.

![caption with only text, web image](http://imgs.xkcd.com/comics/wikipedian_protester.png {width: "500px"})

![caption with only text, web image](http://imgs.xkcd.com/comics/wikipedian_protester.png){width: "500px"}

![caption with only text, web image](http://imgs.xkcd.com/comics/wikipedian_protester.png){width:"500px"}

I have downloaded the image. 

````
![caption with only text, local image](wikipedian_protester.png)
````

It still comes out too small. Nevermind.

It looks as if specifying image size is an ongoing issue, with a complex solution which has been implemented, but not hasn't been fully merged into the general pandoc. Due soon. 
[Syntax for specifying image size #261](https://github.com/jgm/pandoc/issues/261)

[this comment indicates I should try different latex engines.](https://github.com/jgm/pandoc/issues/261#issuecomment-115439719)

Current work around is to amend the image file to specify size in the file header.

All the above were generated running command line:

`````
pandoc CaptionTest5xkcd1.txt --bibliography=xkcd.bib -o CaptionTest5xkcd1.pdf
`````

end of CaptionTest5xkcd1.txt

The next step of testing is to include citations in the caption. 

![caption with @_xkcd citation, web image](http://imgs.xkcd.com/comics/wikipedian_protester.png)

That seems to work.

This doesn't work:

``````
![caption with citation as footnote^[@_xkcd citation] inside caption, web image](http://imgs.xkcd.com/comics/wikipedian_protester.png)

produces error: 

! Argument of \Hy@tempa has an extra }.
<inserted text> 
                \par 
l.161 ...015) citation} inside caption, web image}

pandoc: Error producing PDF from TeX source
`````````

[It looks as if this has already been reported as a bug.](https://github.com/jgm/pandoc/issues/1506)

[This article argues that one should just not have a footnote inside a picture caption.](http://www.tex.ac.uk/FAQ-ftncapt.html)

So, what if I put the footnote after the caption and image. 

![caption with citation as footnote after image link, web image](http://imgs.xkcd.com/comics/wikipedian_protester.png)^[@_xkcd citation] 

Then the caption doesn't appear.

The above tested with this command line:

`````
pandoc CaptionTest5xkcd3.txt --bibliography=xkcd.bib -o CaptionTest5xkcd3.pdf
`````

end of CaptionTest5xkcd3.txt

OK. So I can have a citation plain inside a caption. I can't have it as a footnote within the caption, or after the image. 
All this without specifying the citation style. Let's see if that works now.

All I do in this command is add the csl then pandoc doesn't run.

pandoc CaptionTest5xkcd3.txt --bibliography=xkcd.bib  --csl=../csl/modern-humanities-research-association.csl -o CaptionTest5xkcd4.pdf

results below 

``````
$ pandoc CaptionTest5xkcd4.txt --bibliography=xkcd.bib  --csl=../csl/modern-humanities-research-association.csl -o CaptionTest5xkcd4.pdf

! Argument of \Hy@tempa has an extra }.
<inserted text> 
                \par 
l.153 ...h\footnote{`Xkcd'. } citation, web image}

pandoc: Error producing PDF from TeX source
`````````

I don't know if this is a problem with the citation style file, or with pandoc cite filter.

[I've looked in the citation style repository.](https://github.com/citation-style-language/styles/issues?utf8=%E2%9C%93&q=is%3Aissue+caption)

This post indicates I should bring up the bug on Zotero Forums:
[Make it clear where to discuss a style #40](https://github.com/citation-style-language/styles/issues/40)

[Zotero forum on citation styles](https://forums.zotero.org/11/)

[Apparently Zotero doesn't allow you to put citations in captions.](https://forums.zotero.org/discussion/50392/citations-in-word-figure-and-table-captions-tables/)

To see if it is just the one citation style, I could try different csl files:

``````
$ pandoc CaptionTest5xkcd3.txt --bibliography=xkcd.bib  --csl=../csl/apa.csl -o CaptionTest5xkcd4.pdf
pandoc-citeproc: 9:597-9:606: Expected end element for: Name {nameLocalName = "single", nameNamespace = Nothing, namePrefix = Nothing}, but received: EventEndElement (Name {nameLocalName = "layout", nameNamespace = Nothing, namePrefix = Nothing})
pandoc: Error running filter pandoc-citeproc
Filter returned error status 1
`````

```````
$ pandoc CaptionTest5xkcd3.txt --bibliography=xkcd.bib  --csl=../csl/chicago-annotated-bibliography.csl -o CaptionTest5xkcd4.pdf

! Argument of \Hy@tempa has an extra }.
<inserted text> 
                \par 
l.151 ...ith ibid.\footnote{} citation, web image}

pandoc: Error producing PDF from TeX source
``````

````````````
$ pandoc CaptionTest5xkcd3.txt --bibliography=xkcd.bib  --csl=../csl/chicago-author-date.csl -o CaptionTest5xkcd4.pdf
pandoc-citeproc: ParseError {errorContexts = [], errorMessage = "Failed reading: takeWhile1", errorPosition = 33:5}
pandoc: Error running filter pandoc-citeproc
Filter returned error status 1
`````````````

````````
$ pandoc CaptionTest5xkcd3.txt --bibliography=xkcd.bib  --csl=../csl/chicago-fullnote-bibliography.csl -o CaptionTest5xkcd4.pdf
! Argument of \Hy@tempa has an extra }.
<inserted text> 
                \par 
l.152 ...ith ibid.\footnote{} citation, web image}

pandoc: Error producing PDF from TeX source
`````````

and finally the csl I want to use:

````````
$ pandoc CaptionTest5xkcd3.txt --bibliography=xkcd.bib  --csl=../csl/modern-humanities-research-association.csl -o CaptionTest5xkcd4.pdf
! Argument of \Hy@tempa has an extra }.
<inserted text> 
                \par 
l.153 ...h\footnote{`Xkcd'. } citation, web image}

pandoc: Error producing PDF from TeX source
```````````

OK. So the end result of this testing is that 
a) if I am happy with the built in citation style, then I can only have citations inside captions, not as footnotes within the caption or footnotes after the caption. 
b) if I need non-standard citation style, then I can't even have citations inside a caption. The citation will have to be in the surrounding text. Or, of course, the citation can be typed into the caption. 

a) 

`````
pandoc CaptionTest5xkcd4.txt --bibliography=xkcd.bib -o CaptionTest5xkcd4.pdf
`````
The above commandline doesn't work, seems to choke on what it thinks is a citation tempa.
So, it is probably trying to format what I thought were in code blocks.
I will take those out and make the next version of this document.

**Comment by jgm** [comment]
The reason using the `modern-humanities-research-association.csl` CSL causes problems is that this is a footnote style -- so pandoc-citeproc puts a footnote inside the citation, which as you've already seen does not work (jgm/pandoc#1506).  I'll see if I can make progress on that.  For image sizing, create a larger image or set the DPI in the image.  Image size syntax is coming.

**Comment by colcord** [comment]
Thanks for your prompt response.

The current error messages are quite bewildering to pandoc neophytes.

A more helpful error message from Pandoc, when it comes across a citation
in a caption in a footnote style, would enable a neophyte to identify where
the error was happening. Could the error point to this bug report so they
would learn that they need to change the location of the citation away from
the caption?

I located the problem by searching the text file for text reported in the
pandoc errors. It wasn't clear that the error was caused in a caption or
citation.

thanks again for working on pandoc.


On 1 November 2015 at 21:36, John MacFarlane <notifications@github.com>
wrote:

> The reason using the modern-humanities-research-association.csl CSL
> causes problems is that this is a footnote style -- so pandoc-citeproc puts
> a footnote inside the citation, which as you've already seen does not work (
> jgm/pandoc#1506 <https://github.com/jgm/pandoc/issues/1506>). I'll see if
> I can make progress on that. For image sizing, create a larger image or set
> the DPI in the image. Image size syntax is coming.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jgm/pandoc-citeproc/issues/188#issuecomment-152867166>
> .
>



-- 
frank@colcord.net, cell: +447763116793, skype: colcord,
if you ever get no response, please send to: fcolcord@gmail.com

**Comment by jgm** [comment]
Well, with the change I just committed to pandoc, there
won't be an error any more (though you still won't get a
footnote).

+++ colcord [Nov 01 15 13:59 ]:

**Comment by jgm** [comment]
Resolved by jgm/pandoc#1506

**Comment by colcord** [comment]
Thank you so much for positively resolving the issue so promptly.

I am very grateful.
Frank





On 1 November 2015 at 23:32, John MacFarlane <notifications@github.com>
wrote:

> Resolved by jgm/pandoc#1506 <https://github.com/jgm/pandoc/issues/1506>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jgm/pandoc-citeproc/issues/188#issuecomment-152876522>
> .
>



-- 
frank@colcord.net, cell: +447763116793, skype: colcord,
if you ever get no response, please send to: fcolcord@gmail.com
